src/test/fuzz/util.h

```diff
@@ -642,6 +664,49 @@ class FuzzedSock : public Sock
         return random_bytes.size();
     }
 
+    int Connect(const sockaddr*, socklen_t) const override
+    {
+        // Have a permanent error at connect_errnos[0] because when the fuzzed data is exhausted
+        // SetFuzzedErrNo() will always return the first element and we want to avoid Connect()
+        // returning -1 and setting errno to EAGAIN repeatedly.
+        constexpr std::array connect_errnos{
+            ECONNREFUSED,
+            EAGAIN,
+            ECONNRESET,
+            EHOSTUNREACH,
+            EINPROGRESS,
+            EINTR,
+            ENETUNREACH,
+            ETIMEDOUT,
+        };
+        if (m_fuzzed_data_provider.ConsumeBool()) {
+            SetFuzzedErrNo(m_fuzzed_data_provider, connect_errnos);
+            return -1;
+        }
+        return 0;
+    }
+
+    int Getsockopt(int level, int opt_name, void* opt_val, socklen_t* opt_len) const override
+    {
+        constexpr std::array getsockopt_errnos{
+            ENOMEM,
+            ENOBUFS,
+        };
+        if (m_fuzzed_data_provider.ConsumeBool()) {
+            SetFuzzedErrNo(m_fuzzed_data_provider, getsockopt_errnos);
+            return -1;
+        }
+        if (opt_val == nullptr) {
+            return 0;
+        }
+        const auto rnd = m_fuzzed_data_provider.ConsumeBytes<uint8_t>(*opt_len);
+        std::memcpy(opt_val, rnd.data(), rnd.size());
```

Comment:
Builds cleanly at a6c20d2b88ea92, but I'm seeing a UBSan warning when starting the fuzzer on each of 3 times:

 ```
$ clang --version
clang version 9.0.1-16 
Target: x86_64-pc-linux-gnu
Thread model: posix

$ FUZZ=i2p src/test/fuzz/fuzz ../qa-assets/fuzz_seed_corpus/
INFO: Seed: 3673367385
INFO: Loaded 1 modules   (642733 inline 8-bit counters): 642733 [0x55abe383e4c8, 0x55abe38db375), 
INFO: Loaded 1 PC tables (642733 PCs): 642733 [0x55abe38db378,0x55abe42a9e48), 
INFO:   132706 files found in ../qa-assets/fuzz_seed_corpus/
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 1048576 bytes
INFO: seed corpus: files: 132706 min: 1b max: 3986616b total: 1882953565b rss: 284Mb
#4096	pulse  cov: 1773 ft: 2176 corp: 11/47b exec/s: 2048 rss: 366Mb

test/fuzz/util.h:703:30: runtime error: null pointer passed as argument 2, which is declared to never be null
/usr/include/string.h:44:28: note: nonnull attribute specified here
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior test/fuzz/util.h:703:30 in 

#8192	pulse  cov: 2600 ft: 3813 corp: 32/219b exec/s: 2730 rss: 441Mb
#16384	pulse  cov: 2655 ft: 4890 corp: 70/750b exec/s: 2730 rss: 586Mb
#32768	pulse  cov: 2773 ft: 6429 corp: 111/2011b exec/s: 2730 rss: 649Mb
#65536	pulse  cov: 2794 ft: 7344 corp: 142/4922b exec/s: 2849 rss: 649Mb
#131072	pulse  cov: 2890 ft: 8189 corp: 162/47Kb exec/s: 2473 rss: 1271Mb
#132707	INITED cov: 2890 ft: 8189 corp: 162/47Kb exec/s: 2249 rss: 1271Mb
#132896	REDUCE cov: 2890 ft: 8189 corp: 162/46Kb lim: 14309 exec/s: 2214 rss: 1271Mb L: 13800/13800 MS: 4 ChangeBinInt-ChangeByte-ChangeByte-EraseBytes-
.../...
#204032	REDUCE cov: 2912 ft: 9195 corp: 194/29Kb lim: 14413 exec/s: 2147 rss: 1271Mb L: 62/2551 MS: 2 ChangeBinInt-EraseBytes-
#206627	REDUCE cov: 2912 ft: 9195 corp: 194/29Kb lim: 14426 exec/s: 2152 rss: 1271Mb L: 66/2551 MS: 5 InsertByte-PersAutoDict-ChangeByte-EraseBytes-PersAutoDict- DE: "\x00\x00\x00\x00\x00\x00\x00\x1a"-"\x1a\x00\x00\x00\x00\x00\x00\x00"-
#207638	REDUCE cov: 2912 ft: 9195 corp: 194/29Kb lim: 14426 exec/s: 2162 rss: 1271Mb L: 102/2551 MS: 1 EraseBytes-
```

Reply:
Fixed, thanks! I did not see this because my `memcpy()` does not have the nonnull attribute.

Off-topic: the nonnull attribute has this speciality - the compiler assumes that the parameter will never be null and may eliminate branches in the function that check for that. For example:

```cpp
void func(char* buf) __nonnull(1)
{
    if (buf == nullptr) {
        // take action
        return;
    }
    buf[0] = 1;
}
```

Compiler optimization may reduce this to:

```cpp
void func(char* buf) __nonnull(1)
{
    buf[0] = 1;
}
```

It looks reasonable, given that code like `func(nullptr);` will not compile. However, if the value of the argument is determined at runtime and may still end up being `nullptr`, then the compiler has no way to detect this during compilation. So the line `buf[0] = 1;` may still be executed with `buf` being `nullptr` even though the source code checks for that just above, leading to some nice "wtf" moments.

Ubsan alleviates this.

Reply:
Yes. Runs without the error now.

Reply:
Oops, getting another one.